### DISCUSSION TITLE
Using Qt backend while the app depends on Qt too

### DISCUSSION THREAD

**Discussion by ubruhin** [slint-ui/slint#6352] [General] [answered]
Hi,

I'm trying to understand how Qt is pulled in through Slint and how this might affect applications directly depending on Qt too. Though on Linux I could manage to get it running fine, I have some concerns about deployment (and also experienced already a deployment problem on macOS which *might* be related).

I'm having a C++ application heavily depending on Qt and want to implement the UI with Slint while the C++ business logic still depends on Qt (QString, QThread, QNetworkManager, QEverything). So basically the application now depends on Qt directly and indirectly. But I don't understand yet where Slint gets Qt from. Is it shipped with Cargo? What version does it use? I'm not experienced with Rust yet so at a quick look I didn't see any Qt dependency of `i-slint-backend-qt`.

So, is Qt shipped by Slint separately or does it access the system installed Qt?

```
             ┌─────────────────────┐                               ┌────────────────────┐        
             │ MyApplication (C++) │                               │ MyApplication (C++)│        
             └───┬───────────────┬─┘                               └──┬─────────────┬───┘        
                 │               │                                    │             │            
                 │               │                                    │             │            
                 ▼               ▼                                    │             ▼            
┌──────────────────┐           ┌─────────────────────┐                │   ┌─────────────────────┐
│ Qt (e.g. v6.6.2) │           │ Slint (e.g. v1.8.0) │                │   │ Slint (e.g. v1.8.0) │
└──────────────────┘           └───────────┬─────────┘                │   └─────────┬───────────┘
                                           │                          │             │            
                                           ▼                          │             │            
                               ┌─────────────────────┐                │             │            
                               │ Qt (which version?) │                ▼             ▼            
                               └─────────────────────┘             ┌──────────────────┐          
                                                                   │ Qt (e.g. v6.6.2) │          
                                                                   └──────────────────┘          
```

I think if the app and Slint both access the same Qt, everything is good and deployment *should* not cause issues. Could you maybe give a hint how the Qt dependency is pulled in? If it uses the system installed Qt, what minimum version is required?

And could you confirm the official Slint binaries do *not* contain the Qt backend? I couldn't find any documentation about this.

Due to the dependency to Qt, we can't use any other Slint backend. We require the Qt event loop to be running and as far as I understand, that's only possible by using the Qt backend.

Thanks!

**Comment by ogoffart** [comment, marked as answer]
Ideally it should use the same Qt. Anything else is asking for troubles.
When compiled using cmake, this code is used to find Qt:

https://github.com/slint-ui/slint/blob/2c9e79613e142c4a68aaebe3bd58621e385e6386/api/cpp/CMakeLists.txt#L210-L226

(That code sets the QMAKE env variable to build the rust integration, so it will use that version of Qt)

> And could you confirm the official Slint binaries do not contain the Qt backend? I couldn't find any documentation about this.

The windows and mac one are compiled only with the winit backend.
The Linux ones don't contain the Qt backend and assume Qt 5.15 is present on the system, IIRC

> Due to the dependency to Qt, we can't use any other Slint backend. We require the Qt event loop to be running and as far as I understand, that's only possible by using the Qt backend.

Right, I would recommend using the Qt backend in your case. 
An alternative (not necessarily recommended) is to create your own backend in C++ using Qt like done there:  https://github.com/slint-ui/slint/blob/master/examples/cpp/platform_qt/main.cpp

**Reply by ubruhin** [reply]
>  When compiled using cmake, this code is used to find Qt:

Ah, perfect! So Qt 6.2 (or even 5.15) is currently required, which is great (nothing too bleeding-edge).

> The windows and mac one are compiled only with the winit backend.
> The Linux ones don't contain the Qt backend and assume Qt 5.15 is present on the system, IIRC

Alright, so on Windows & mac I'll have to compile it from sources.

Thanks for the clarifications!
